traffic_ops/traffic_ops_golang/user/current.go

```diff
@@ -80,3 +190,197 @@ WHERE u.id=$1
+		hashPass, err := auth.DerivePassword(*user.LocalPassword)
+		if err != nil {
+			sysErr = fmt.Errorf("Hashing new password: %v", err)
+			errCode = http.StatusInternalServerError
+			api.HandleErr(w, r, tx, errCode, nil, sysErr)
+			return
+		}
+		changePasswd = true
+		user.LocalPassword = util.StrPtr(hashPass)
+		user.ConfirmLocalPassword = util.StrPtr(hashPass)
+	}
+
+	if *user.Role != inf.User.Role {
+		privLevel, exists, err := dbhelpers.GetPrivLevelFromRoleID(tx, *user.Role)
+		if err != nil {
+			sysErr = fmt.Errorf("Getting privLevel for Role #%d: %v", *user.Role, err)
+			errCode = http.StatusInternalServerError
+			api.HandleErr(w, r, tx, errCode, nil, sysErr)
+			return
+		}
+		if !exists {
+			userErr = fmt.Errorf("role: no such role: %d", *user.Role)
+			errCode = http.StatusNotFound
+			api.HandleErr(w, r, tx, errCode, userErr, nil)
+			return
+		}
+		if privLevel > inf.User.PrivLevel {
+			userErr = errors.New("role: cannot have greater permissions than user's current role")
+			errCode = http.StatusForbidden
+			api.HandleErr(w, r, tx, errCode, userErr, nil)
+			return
+		}
+	}
+
+	if ok, err := tenant.IsResourceAuthorizedToUserTx(*user.TenantID, inf.User, tx); err != nil {
+		if err == sql.ErrNoRows {
+			userErr = errors.New("No such tenant!")
+			errCode = http.StatusNotFound
+		} else {
+			sysErr = fmt.Errorf("Checking user %s permissions on tenant #%d: %v", inf.User.UserName, *user.TenantID, err)
+			errCode = http.StatusInternalServerError
+		}
+		api.HandleErr(w, r, tx, errCode, userErr, sysErr)
+		return
+	} else if !ok {
+		// unlike Perl, this endpoint will not disclose the existence of tenants over which the current
+		// user has no permission - in keeping with the behavior of the '/tenants' endpoint.
+		userErr = errors.New("No such tenant!")
+		errCode = http.StatusNotFound
+		api.HandleErr(w, r, tx, errCode, userErr, sysErr)
+		return
+	}
+
+	if *user.Username != inf.User.UserName {
+
+		if ok, err := dbhelpers.UsernameExists(*user.Username, tx); err != nil {
+			sysErr = fmt.Errorf("Checking existence of user %s: %v", *user.Username, err)
+			errCode = http.StatusInternalServerError
+			api.HandleErr(w, r, tx, errCode, nil, sysErr)
+			return
+		} else if ok {
+			// TODO users are tenanted, so theoretically I should be hiding the existence of the
+			// conflicting user - but then how do I tell the client how to fix their request?
+			userErr = fmt.Errorf("Username %s already exists!", *user.Username)
+			errCode = http.StatusConflict
+			api.HandleErr(w, r, tx, errCode, userErr, nil)
+			return
+		}
+	}
+
+	if err = updateUser(&user, tx, changePasswd); err != nil {
+		errCode = http.StatusInternalServerError
+		sysErr = fmt.Errorf("Updating user: %v", err)
+		api.HandleErr(w, r, tx, errCode, nil, sysErr)
+		return
+	}
+
+	api.WriteRespAlertObj(w, r, tc.SuccessLevel, "User profile was successfully updated", user)
+}
+
+func updateUser(u *tc.User, tx *sql.Tx, cp bool) error {
+	var row *sql.Row
+	if cp {
```

Comment:
It is not clear why two different queries are needed here. Can you explain?

Reply:
Because if you don't change your password, then `u.LocalPassword` and `u.ConfirmLocalPassword` are both either `nil` or pointers to empty strings. It would be unacceptable to update a user to have either of those values as a password. So I update those fields if and only if a password change was requested.

Reply:
I think I see, so there is some UI on the client side that only passes password data if the user opted in to changing it, and `GetUserByID` does not populate the password fields so we can't just write those back. Maybe a comment explaining these issues would help avoid mistakes by future maintainers.

Reply:
You mean on the the `updateUser` subroutine or like a warning on `GetUserByID` stating that it doesn't return password info?

Reply:
I was thinking on the `updateUser` function. I have a different suggestion that occurred to me over the weekend, though.

Would it be acceptable to split the updating of the password fields and the rest of the user fields into two separate queries?

Since `updateUser` is given a `sql.Tx`, I assume that the caller has proper transaction management with rollback if we return an error so that two queries would have the same atomicity properties as the current code. It does have the potential to be slightly less performant, but I don't know how much that matters here.

On the other hand, I think we would gain a lot of code simplicity and readability. At the moment we have an if statement that chooses between two almost identical queries that are long enough to make it difficult to eyeball the difference between them. But if we split out the password update query we end up with an `updateUser` function structured along the lines of this:

```go
func updateUser(u *tc.User, tx *sql.Tx, cp bool) error {
    row := tx.QueryRow(replaceCurrentQuery,
        u.AddressLine1,
        ...
    )
    err := row.Scan(&u.AddressLine1,
        ...
    )
    if err != nil {
        return err
    }

    if cp {
        _, err := tx.Exec(replacePasswordQuery, u.LocalPassword, u.ConfirmLocalPassword)
        if err != nil {
            return err
        }
        u.LocalPassword = nil
        u.ConfirmLocalPassword = nil
    }
    return nil
}
```

(Note: I'm not sure about the correct place to nil out the password fields here, so adjust as necessary.)

With the above structure it becomes much easier to see what the `cp` parameter controls and how the password fields are managed. I think it is also less likely for a future maintainer to screw up the password handling.

Thoughts?